### PR TITLE
Add Terminal Locations REST API

### LINKED ISSUE #2784: Terminal Locations API: endpoint to retrieve store location
Relying on store address, specified in WC setting, interact with server endpoints:
- get existing locations
- create a new one if the is no one matching the settings
- return the location details as response

Caching/using transients would be a great thing to have for better performance and reliability.

Route: `terminals/locations/store`
Inspiration for the new controller: `\WC_REST_Payments_Accounts_Controller` + its tests.

Depends on 1103-gh-Automattic/woocommerce-payments-server and 1107-gh-Automattic/woocommerce-payments-server

### PR DESCRIPTION
**Pull request by RadoslavGeorgiev**
Fixes #2784
Requires 1145-gh-Automattic/woocommerce-payments-server

#### Changes proposed in this Pull Request

A new `terminal/locations/store` endpoint, which utilizes the `/wcpay/terminal/locations` GET and POST server endoints.

This endpoint does not simply forward the request to Stripe, instead it ensures that a correct singular location is always return based on the settings of the particular shop.

To accomplish this, the store settings are prepared upon each call, and an existing location is returned only whenever it completely matches them. If any of the following settings changes, this will require a new location to be created:

- Blog/store name
- City
- Country
- Address line 1
- Address line 2
- Postal code
- State

A transient is used to cache all existing locations for a day, in order to avoid redundant API calls. Note that only the existing locations are cached, not only the one for the store. This ensures that if settings get changed, a new location will be generated, without using excessive resources.

image

#### Testing instructions

0. Run unit tests and make sure that they are sound.
1. Make sure your WCPay account (in the local server) has the `card_present_enabled` meta set to `'1'`.
2. Call the new endpoint and make sure that the response contains the correct data for the current store.
3. Call the endpoint again and verify that the same location is returned, instead of generating a new one.
4. (Optionally) Clear all transients and call the endpoint once more.

-------------------

- [ ] Added changelog entry: **Does a changelog entry make sense for this PR, considering that it only makes sense for merchants who are eligible for card present?**
- [x] Covered with tests
- [x] Tested on mobile (does not apply)

**Post merge**

- [ ] Added testing instructions to the [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) (or does not apply)

### CODE REVIEW

**Review by htdat — CHANGES_REQUESTED** [review]
Great work! 

✅ Followed up test steps and worked well. 
✅ Cleared up the transient, tried to change the store name and address, got expected results. 

I've left a few inline comments to address some small issues. Other than that, things look good.

File: includes/wc-payment-api/class-wc-payments-api-client.php [hunk 1]
```diff
@@ -1148,6 +1149,60 @@ public function register_domain_with_apple( $domain_name ) {
...
+	 * @see https://stripe.com/docs/api/terminal/locations/object
+	 *
+	 * @throws API_Exception If an error occurs.
+	 */
+	public function create_terminal_location( $display_name, $address, $metadata = null ) {
+		if ( ! is_array( $address ) || ! isset( $address['country'] ) || ! isset( $address['line1'] ) ) {
+			throw new API_Exception(
+				__( 'Address country and line 1 are required.', 'woocommerce-payments' ),
+				'wcpay_invalid_terminal_location_request',
+				0
```
[the remark this answers is not in the document]
Comment (RadoslavGeorgiev) [reply includes/wc-payment-api/class-wc-payments-api-client.php]: It should be exactly 400. I was looking at another manual exception somewhere else in the plugin, and blindly copied the 0 from it. Fixed in https://github.com/Automattic/woocommerce-payments/pull/2830/commits/8eeca56f2dbff6a242abbbb9496b0ec71d61e8ac

File: includes/wc-payment-api/class-wc-payments-api-client.php [hunk 2]
```diff
@@ -1148,6 +1149,60 @@ public function register_domain_with_apple( $domain_name ) {
...
+			);
+		}
+
+		$request = [
+			'display_name' => $display_name,
+			'address'      => $address,
+		];
```
[the remark this answers is not in the document]
Comment (RadoslavGeorgiev) [reply includes/wc-payment-api/class-wc-payments-api-client.php]: Makes sense. In https://github.com/Automattic/woocommerce-payments/pull/2830/commits/8eeca56f2dbff6a242abbbb9496b0ec71d61e8ac I changed the default value to an empty array, and added it to the request.

File: includes/wc-payment-api/class-wc-payments-api-client.php [hunk 3]
```diff
@@ -1148,6 +1149,60 @@ public function register_domain_with_apple( $domain_name ) {
...
+	 * }
+	 * @param mixed[] $metadata Optional. Metadata for the location.
+	 *
+	 * @return array A Stripe terminal location object.
+	 * @see https://stripe.com/docs/api/terminal/locations/object
+	 *
+	 * @throws API_Exception If an error occurs.
+	 */
+	public function create_terminal_location( $display_name, $address, $metadata = null ) {
+		if ( ! is_array( $address ) || ! isset( $address['country'] ) || ! isset( $address['line1'] ) ) {
```
[the remark this answers is not in the document]
Comment (RadoslavGeorgiev) [reply includes/wc-payment-api/class-wc-payments-api-client.php]: That's a cleaner syntax, added in https://github.com/Automattic/woocommerce-payments/pull/2830/commits/8eeca56f2dbff6a242abbbb9496b0ec71d61e8ac 👌

**Comment by RadoslavGeorgiev** [comment]
Hey @htdat,

I've addressed your comments! Could you please take another look?

**Review by htdat — APPROVED** [review]
> I've addressed your comments! Could you please take another look?

Thank you! 🚢 LGTM now.

**Comment by vbelolapotkov** [comment]
Hey @RadoslavGeorgiev ! I've noticed some e2e tests are failing. Do you know the reason? I know sometimes they are fragile, but worth double checking before merging to avoid breaking some flows.

**Comment by RadoslavGeorgiev** [comment]
@vbelolapotkov I'm not sure why that's happening. This is a completely separate endpoint, which should not affect E2E tests in any way, and I mostly see time-outs. Out of curiosity, I re-ran the tests and I'm interested to see the result in a few minutes.

**Comment by RadoslavGeorgiev** [comment]
It worked after re-running the tests, **and** after bringing it up to date with `develop`, without timeouts. I think it should be fine 🤞

**Pull request merged**